kamaki/cli/cmds/cyclades.py

```diff
@@ -1145,3 +1148,152 @@ def main(self, server_id):
                     self.arguments['attachment_id'].lvalue,
                     self.arguments['volume_id'].lvalue)])
         self._run(server_id=server_id)
+
+
+@command(keypair_cmds)
+class keypair_list(_CycladesInit, OptionalOutput, NameFilter):
+    """List all keypairs"""
+
+    arguments = dict(
+        detail=FlagArgument('show detailed output', ('-l', '--detail')),
+    )
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    def _run(self):
+        keypairs = self.client.list_keypairs()
+        keypairs = [k['keypair'] for k in keypairs]
+        if not self['detail']:
+            remove_from_items(keypairs, 'public_key')
+        keypairs = self._filter_by_name(keypairs)
+        self.print_(keypairs)
+
+    def main(self):
+        super(self.__class__, self)._run()
+        self._run()
+
+
+@command(keypair_cmds)
+class keypair_info(_CycladesInit, OptionalOutput):
+    """Detailed information on a keypair"""
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    @errors.Keypair.name
+    def _run(self, key_name):
+        keypair = self.client.get_keypair_details(key_name)
+        self.print_(keypair, self.print_dict)
+
+    def main(self, key_name):
+        super(self.__class__, self)._run()
+        self._run(key_name=key_name)
+
+
+@command(keypair_cmds)
+class keypair_delete(_CycladesInit):
+    """Delete a keypair"""
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    @errors.Keypair.name
+    def _run(self, key_name):
+        keypair = self.client.delete_keypair(key_name)
+
+    def main(self, key_name):
+        super(self.__class__, self)._run()
+        self._run(key_name=key_name)
+
+
+@command(keypair_cmds)
+class keypair_upload(_CycladesInit, OptionalOutput):
+    """Upload or update a keypair"""
+
+    arguments = dict(
+        key_name=ValueArgument(
+            'The name of the key',
+            '--key-name'),
+        public_key=ValueArgument(
+            'The contents of the keypair(the public key)',
+            '--public-key'),
+        force=FlagArgument(
+            'Force key name override', ('-f', '--force')
+        )
+    )
+
+    required = ['public_key']
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    def _run(self):
+
+        if self['key_name'] is not None:
+            active_keypairs = self.client.list_keypairs()
```

Comment:
Again, keypairs are only used to check if a specific key_name is included, therefore we should rather generate the `active_keynames` list and check if the key_name is included:
```python
active_keynames = [k['keypair']['name'] for k in self.client.list_keypais()]
if self['key_name'] in active_keynames and not self['force']:
    raise CLIError(...
```